### PR TITLE
Add goQuorum node

### PR DESCRIPTION
**Pull request by rain-on**

### CODE REVIEW

File: dsl/src/main/java/tech/pegasys/peeps/network/Network.java [hunk 1]
```diff
@@ -173,25 +189,30 @@ public Besu addNode(
...
-                .withContainerNetwork(subnet.network())
-                .withIpAddress(subnet.getAddressAndIncrement())
-                .withGenesisFile(genesisFile)
-                .withBootnodeEnodeAddress(bootnodeEnodeAddresses())
-                .build());
+  private Web3Provider addNode(
+      final Web3ProviderConfigurationBuilder config, final Web3ProviderType providerType) {
+    final Web3Provider web3Provider;
+    config
+        .withVertx(vertx)
```
Comment (usmansaleem) [inline dsl/src/main/java/tech/pegasys/peeps/network/Network.java:204]: do we need to set this if we are building `GoQuorum` ? or it will simply be ignored by GoQorum constructor?

File: dsl/src/main/java/tech/pegasys/peeps/node/GoQuorum.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        MountableFile.forHostPath(config.getGenesisFile()), CONTAINER_GENESIS_FILE);
+    final List<String> entryPoint = Lists.newArrayList("/bin/sh", "-c");
+    final String initCmd =
+        "mkdir -p '"
+            + DATA_DIR
+            + "/geth' && geth --datadir \""
+            + DATA_DIR
+            + "\" init "
+            + CONTAINER_GENESIS_FILE
+            + " && echo '##### GoQuorum INITIAILSED #####' && ";
```
Comment (usmansaleem) [inline dsl/src/main/java/tech/pegasys/peeps/node/GoQuorum.java:58]: nit: spelling

**Review by usmansaleem — APPROVED** [review]
Looks OK to me.

File: dsl/src/main/java/tech/pegasys/peeps/node/GoQuorum.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+    addNodePrivateKey(config, commandLineOptions, container);
+    //    if (config.isPrivacyEnabled()) {
```
Comment (jframe) [inline dsl/src/main/java/tech/pegasys/peeps/node/GoQuorum.java:61]: Will the privacy config for Quorum be in another PR?

File: dsl/src/main/java/tech/pegasys/peeps/node/GoQuorum.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    return Wait.forLogMessage(".*endpoint=0.0.0.0:8545.*", 1);
+  }
+
+  private List<String> standardCommandLineOptions() {
+    return Lists.newArrayList(
+        "--verbosity",
+        "5",
+        "--rpccorsdomain",
+        "\"*\"",
+        // "--mine",
```
Comment (jframe) [inline dsl/src/main/java/tech/pegasys/peeps/node/GoQuorum.java:90]: Would be good for GoQuorum to be configured as a miner same as the Besu node

File: dsl/src/main/java/tech/pegasys/peeps/node/GoQuorum.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+      final List<String> commandLineOptions,
+      final GenericContainer<?> container) {
+    container.withClasspathResourceMapping(
+        config.getNodeKeyPrivateKeyResource().get(),
+        CONTAINER_NODE_PRIVATE_KEY_FILE,
+        BindMode.READ_ONLY);
+    commandLineOptions.addAll(Lists.newArrayList("--nodekey", CONTAINER_NODE_PRIVATE_KEY_FILE));
+  }
+
+  //  private void addPrivacy(
```
Comment (jframe) [inline dsl/src/main/java/tech/pegasys/peeps/node/GoQuorum.java:139]: Same as above, assuming this can be removed until Privacy config is added in?

File: dsl/src/main/java/tech/pegasys/peeps/node/genesis/ibft2/BftConfig.java [hunk 6]
```diff
@@ -14,17 +14,18 @@
 
 import com.fasterxml.jackson.annotation.JsonGetter;
 
-public class Ibft2Config {
+// NOTE: This is only for Besu (GoQuorum needs their own version
+public class BftConfig {
```
Comment (jframe) [inline dsl/src/main/java/tech/pegasys/peeps/node/genesis/ibft2/BftConfig.java:18]: Think this should be under a bft package now

File: end-to-end-tests/src/test/java/tech/pegasys/peeps/consensus/GoQuorumCliqueConsensusTest.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+import tech.pegasys.peeps.node.Web3ProviderType;
+import tech.pegasys.peeps.node.model.Hash;
+import tech.pegasys.peeps.node.verification.ValueReceived;
+import tech.pegasys.peeps.node.verification.ValueSent;
+
+import org.apache.tuweni.eth.Address;
+import org.apache.tuweni.units.ethereum.Wei;
+import org.junit.jupiter.api.Test;
+
+public class GoQuorumCliqueConsensusTest extends NetworkTest {
```
Comment (jframe) [inline end-to-end-tests/src/test/java/tech/pegasys/peeps/consensus/GoQuorumCliqueConsensusTest.java:30]: Would be possible to parameterise this test for Besu or GoQuorum. This and the CliqueConsensusTest almost exactly the same except the line 37.

File: end-to-end-tests/src/test/java/tech/pegasys/peeps/consensus/IbftConsensusTest.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+import tech.pegasys.peeps.network.ConsensusMechanism;
+import tech.pegasys.peeps.network.Network;
+
+public class IbftConsensusTest extends NetworkTest {
+
+  private final NodeConfiguration node = NodeConfiguration.ALPHA;
+  private final SignerConfiguration signer = SignerConfiguration.ALPHA;
+
+  @Override
+  protected void setUpNetwork(final Network network) {
```
Comment (jframe) [inline end-to-end-tests/src/test/java/tech/pegasys/peeps/consensus/IbftConsensusTest.java:27]: Can you add a test for the IBFT network? This test doesn't currently do anything.

File: dsl/src/main/java/tech/pegasys/peeps/node/Web3Provider.java [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+        "enode://%s@%s:%d", pubKey, config.getIpAddress().get(), CONTAINER_P2P_PORT);
+  }
+
+  private String nodePublicKey(final Web3ProviderConfiguration config) {
+    return removeAnyHexPrefix(ClasspathResources.read(config.getNodeKeyPublicKeyResource().get()));
+  }
+
+  private void logPortMappings() {
+    LOG.info(
+        "Besu Container: {}, HTTP RPC port mapping: {} -> {}, WS RPC port mapping: {} -> {}, p2p port mapping: {} -> {}",
```
Comment (macfarla) [inline dsl/src/main/java/tech/pegasys/peeps/node/Web3Provider.java:211]: s/Besu/Web3Provider

File: dsl/src/main/java/tech/pegasys/peeps/node/Web3Provider.java [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+        container.getMappedPort(CONTAINER_HTTP_RPC_PORT),
+        CONTAINER_WS_RPC_PORT,
+        container.getMappedPort(CONTAINER_WS_RPC_PORT),
+        CONTAINER_P2P_PORT,
+        container.getMappedPort(CONTAINER_P2P_PORT));
+  }
+
+  private void logContainerNetworkDetails() {
+    if (container.getNetwork() == null) {
+      LOG.info("GoQuorum Container: {}, has no network", container.getContainerId());
```
Comment (macfarla) [inline dsl/src/main/java/tech/pegasys/peeps/node/Web3Provider.java:223]: s/GoQuorum/Web3Provider

File: dsl/src/main/java/tech/pegasys/peeps/node/Web3Provider.java [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    } else {
+      LOG.info(
+          "GoQuorum Container: {}, IP address: {}, Network: {}",
```
Comment (macfarla) [inline dsl/src/main/java/tech/pegasys/peeps/node/Web3Provider.java:226]: s/GoQuorum/Web3Provider

File: dsl/src/main/java/tech/pegasys/peeps/network/Network.java (hunk 1 above)
Comment (rain-on) [reply dsl/src/main/java/tech/pegasys/peeps/network/Network.java:204]: The Vertx is how the test frameworks talks to the node, i.e. it doesn't get passed _into_ the node.

File: dsl/src/main/java/tech/pegasys/peeps/node/GoQuorum.java (hunk 2 above)
Comment (rain-on) [reply dsl/src/main/java/tech/pegasys/peeps/node/GoQuorum.java:59]: done

File: dsl/src/main/java/tech/pegasys/peeps/network/Network.java [hunk 12]
```diff
@@ -155,14 +158,27 @@ public void set(final ConsensusMechanism consensus, final Besu... validators) {
             consensus, Account.of(Account.ALPHA, Account.BETA, Account.GAMMA), validators);
   }
 
-  public Besu addNode(final NodeIdentifier frameworkIdentity, final NodeKey ethereumIdentiity) {
+  public Web3Provider addNode(
+      final NodeIdentifier frameworkIdentity, final NodeKey ethereumIdentiity) {
```
[the remark this answers is not in the document]
Comment (rain-on) [reply dsl/src/main/java/tech/pegasys/peeps/network/Network.java:162]: done

File: dsl/src/main/java/tech/pegasys/peeps/network/Network.java [hunk 13]
```diff
@@ -155,14 +158,27 @@ public void set(final ConsensusMechanism consensus, final Besu... validators) {
...
             .withIdentity(frameworkIdentity)
-            .withNodeKey(ethereumIdentiity));
+            .withNodeKey(ethereumIdentiity),
+        Web3ProviderType.BESU);
   }
 
-  public Besu addNode(
+  public Web3Provider addNode(
+      final NodeIdentifier frameworkIdentity,
+      final NodeKey ethereumIdentiity,
```
[the remark this answers is not in the document]
Comment (rain-on) [reply dsl/src/main/java/tech/pegasys/peeps/network/Network.java:172]: done

File: dsl/src/main/java/tech/pegasys/peeps/network/Network.java [hunk 14]
```diff
@@ -155,14 +158,27 @@ public void set(final ConsensusMechanism consensus, final Besu... validators) {
...
+    return addNode(
+        new Web3ProviderConfigurationBuilder()
+            .withIdentity(frameworkIdentity)
+            .withNodeKey(ethereumIdentiity),
+        providerType);
+  }
+
+  public Web3Provider addNode(
       final NodeIdentifier identity,
       final NodeKey ethereumIdentiity,
```
[the remark this answers is not in the document]
Comment (rain-on) [reply dsl/src/main/java/tech/pegasys/peeps/network/Network.java:183]: oh yes - done.

File: dsl/src/main/java/tech/pegasys/peeps/node/GoQuorum.java (hunk 3 above)
Comment (rain-on) [reply dsl/src/main/java/tech/pegasys/peeps/node/GoQuorum.java:62]: yeah, that's a different piece of the puzzle - this just introduces GoQuorum and makes it work

File: dsl/src/main/java/tech/pegasys/peeps/node/GoQuorum.java (hunk 4 above)
Comment (rain-on) [reply dsl/src/main/java/tech/pegasys/peeps/node/GoQuorum.java]: Yes it would, but that requires Wallets and things, which I think can be moved to the next ticket.

File: dsl/src/main/java/tech/pegasys/peeps/node/GoQuorum.java (hunk 5 above)
Comment (rain-on) [reply dsl/src/main/java/tech/pegasys/peeps/node/GoQuorum.java:147]: _sort of_ like it here as a statement that it needs to be done, but .... can be convinced.

File: end-to-end-tests/src/test/java/tech/pegasys/peeps/consensus/IbftConsensusTest.java (hunk 8 above)
Comment (rain-on) [reply end-to-end-tests/src/test/java/tech/pegasys/peeps/consensus/IbftConsensusTest.java:27]: removed this test - in order for it to work, we need to actually add in IBFT genesis file generation.

File: dsl/src/main/java/tech/pegasys/peeps/node/Web3Provider.java (hunk 9 above)
Comment (rain-on) [reply dsl/src/main/java/tech/pegasys/peeps/node/Web3Provider.java:221]: done

File: dsl/src/main/java/tech/pegasys/peeps/node/Web3Provider.java (hunk 10 above)
Comment (rain-on) [reply dsl/src/main/java/tech/pegasys/peeps/node/Web3Provider.java:233]: done

File: dsl/src/main/java/tech/pegasys/peeps/node/Web3Provider.java (hunk 11 above)
Comment (rain-on) [reply dsl/src/main/java/tech/pegasys/peeps/node/Web3Provider.java:236]: done

File: end-to-end-tests/src/test/java/tech/pegasys/peeps/consensus/GoQuorumCliqueConsensusTest.java (hunk 7 above)
Comment (rain-on) [reply end-to-end-tests/src/test/java/tech/pegasys/peeps/consensus/GoQuorumCliqueConsensusTest.java:30]: yes we should, but I _think_ for now, we get such bang-for-our-buck just having GoQuorum in, that I don't really want to mess around with niceties - i.e. I'm _personally_ happy with this tech debt ... but I can be convinced if really unhappy :)

**Review by jframe — APPROVED** [review]
